# Not-Alike pipeline

## Overview

This pipeline identifies genomic regions which are the most dissimilar compared at least to one genome of a database composed with several genomes sequences. It uses bioinformatic tools to acomplish this task such as: blastn [Download](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/), hisat2 [Download](https://daehwankimlab.github.io/hisat2/download/#version-hisat2-221), samtools [Download](https://www.htslib.org/download/), stringtie [Download](https://ccb.jhu.edu/software/stringtie/#install) and gffread [Download](https://github.com/gpertea/gffread). These executables should be inside a binaries folder and exported to PATH environment variable. This pipeline has three steps: **(1)** split query genome, **(2)** iterative searching and **(3)** mapping & assembling. The result is a GTF file generated by stringtie which assembles the dissimilar fragments not-alike found. Using gffread you can get the sequences of those assembled fragments.

## Introduction

This commands pipeline finds unique dissimilar regions of the genome of interest by mean of interative sequence comparison between all posible fragments of size "window size" at each step size "step size" of the genome. It is written in Python language and uses external bioinformatic tools such as: blastn, hisat2, samtools, stringtie and gffread. Also it uses the following python packages: click, os, random and subprocess that handles the communication with python and the external tools. This commands pipeline works with datasets downloaded from NCBI, thus it is needed to install datasets and dataformat NCBI tools to manipulate genomes database obtained from NCBI datasets. As we pointed out above, it has three steps.

**Split genome step**. In this step the genome of interest is split into fragments of defined size (window size) with a sliding window method that cuts fragments at each step of determined size (step size). In the **iterative searching step**, fragments are used as query to search the genomes database, genome by genome, and fragments that hit a subsequence of the searched genome are elimineated form fragments file. This procedure is performed iterativelly until the end of genomes database. And the final step, the **mapping & assembly**. In this step the remained fragments are subject to be mapped to reference genome, which is the same genome where fragments come from, with hisat2, samtools and stringtie. The resulting GTF file contains the coordinates of the assembled fragments. Those assembled fragments are the dissimilar regions.


## Usage

To ask for help.

```bash
not-alike3 --help

Usage: not-alike3 [OPTIONS] COMMAND [ARGS]...

  Not-Alike: Command pipeline that identifies dissimilar regions of a target
  genome by comparing it to a genomes database.

Options:
  --help  Show this message and exit.

Commands:
  assm-stats      Calculates assembly statistics such as: Mean, Median,...
  db-makeblast    Builds a BLAST_DB (version 5) database files
  db-makefile     Creates the database text file which contains the...
  search          Searches for not alike fragments in query genome
  search-primers  Selects the best fitted primer sequences based on input...
  show-db         Shows metadata of database [accession number, organism...
  show-exp        Shows information about epxeriments stored in the...
```

To ask for help for every command.

```bash
not-alike3 <command> --help
```

Build database. Database path is where FASTA genomes folders and the json files of metadata are.

```bash
not-alike3 db-makeblast -db /path/to/database
```

Prepare the database text file. It contains the sorted path and names of genomes file

```bash
not-alike3 db-makefile -qg query/target_genome.fasta -db /path/to/database/ -e ACCESSION,...,ACCESSION | -i ACCESSION,...,ACCESSION -o database_text_file.txt
```

Search task execution.

```bash
not-alike search -g test/seqs/target_genome.fasta -db /path/to/database/database_text_file.txt -c 'Leave a comment encolsed by single quotes' --config-file <file.toml>
```

Design primers.

```bash
not-alike3 search-primers --input-file unique_dissimilar_regions.fasta --opt-size 20 --opt-gc 55 --opt-tm 58 --product-size 100-500 --template-size-range 1000-2000
```

Example of TOML configuration file

```bash
# Commentaries
[global]
num_of_cores = 2
quite_opposite = false

[[split]]
window_size = 500
step_size = 50

[[search]]
quite_opposite = false
name = 'test2'
task = 'blastn'
identity = 50
qcov = 50
evalue = 100
```

## Commands

### database-makeblastdb

Creates a database of BLAST\_DB version 5 files from a NCBI genomes dataset downloaded with dataset tool. It reads the dataset\_catalog.json to get the path to FASTA files.

### database-makefiledb

Creates a text file with the path and names where BLAST\_DB file are separated by return characters. This command uses a quick sorting prcedure based on sequence comparison to sort database from most similar to the most dissimilar database genome compared with the genome of interest.

### search

This command finds unique dissimilar regions of a genome of interest by means of iterative sequence comparison analysis.

### search-primers

This command design PCR primers over unique dissimilar regions identified by **search** command.

### show-database

This command transforms the assembly\_data\_report.jsonl file, metadata downloaded from NCBI datasets, to TSV format file with dataformat NCBI tool and prints on screen the accession number, organism name and ID of organism taxon.

### show-exp

It reads the 'experiments.log' log file stored in log folder and prints on screen the information contained in it which are information about parameters values, process ID and the date of experiments previously done.

## Download & install


#### Download the zipped file at [exseivier's github](https://www.github.com/exseivier/not-alike3-3.0.0)

```bash
unzip main.zip

cd not-alike3-3.0.0-main

python3 -m pip3 install dist/not-alike3-3.0.0.tar.gz
```

## Wtach the youtube video that explains download, installation and usage instructions.

